server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQAwsS3JmapTestRule.java

```diff
@@ -0,0 +1,97 @@
+import org.apache.james.mailbox.extractor.TextExtractor;
+import org.apache.james.mailbox.store.search.PDFTextExtractor;
+import org.apache.james.modules.TestAwsS3BlobStoreModule;
+import org.apache.james.modules.TestESMetricReporterModule;
+import org.apache.james.modules.TestJMAPServerModule;
+import org.apache.james.modules.TestRabbitMQModule;
+import org.apache.james.server.core.configuration.Configuration;
+import org.junit.rules.TemporaryFolder;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+import com.google.inject.Module;
+
+public class CassandraRabbitMQAwsS3JmapTestRule implements TestRule {
+
+    private static final int LIMIT_TO_10_MESSAGES = 10;
+    public static final int TWO_SECONDS = 2000;
+    private final TemporaryFolder temporaryFolder;
+
+    public static CassandraRabbitMQAwsS3JmapTestRule defaultTestRule() {
+        return new CassandraRabbitMQAwsS3JmapTestRule(new EmbeddedElasticSearchRule());
+    }
+
+    private final GuiceModuleTestRule guiceModuleTestRule;
+
+    public CassandraRabbitMQAwsS3JmapTestRule(GuiceModuleTestRule... guiceModuleTestRule) {
+        TempFilesystemTestRule tempFilesystemTestRule = new TempFilesystemTestRule();
+        this.temporaryFolder = tempFilesystemTestRule.getTemporaryFolder();
+        this.guiceModuleTestRule =
+            AggregateGuiceModuleTestRule
+                .of(guiceModuleTestRule)
+                .aggregate(tempFilesystemTestRule);
+    }
+
+    public GuiceJamesServer jmapServer(Module... additionals) throws IOException {
+        Configuration configuration = Configuration.builder()
+            .workingDirectory(temporaryFolder.newFolder())
+            .configurationFromClasspath()
+            .build();
+
+        return GuiceJamesServer.forConfiguration(configuration)
+            .combineWith(CassandraRabbitMQJamesServerMain.MODULES)
+            .overrideWith(binder -> binder.bind(TextExtractor.class).to(PDFTextExtractor.class))
+            .overrideWith(new TestRabbitMQModule(DockerRabbitMQSingleton.SINGLETON))
+            .overrideWith(new TestAwsS3BlobStoreModule())
+            .overrideWith(new TestJMAPServerModule(LIMIT_TO_10_MESSAGES))
+            .overrideWith(new TestESMetricReporterModule())
+            .overrideWith(guiceModuleTestRule.getModule())
+            .overrideWith((binder -> binder.bind(CleanupTasksPerformer.class).asEagerSingleton()))
+            .overrideWith(additionals);
+    }
+
+    @Override
+    public Statement apply(Statement base, Description description) {
+        return guiceModuleTestRule.apply(base, description);
+    }
+
+    public void await() {
+        awaitProcessingStart();
+        guiceModuleTestRule.await();
+    }
+
+    private void awaitProcessingStart() {
+        // As the RabbitMQEventBus is asynchronous we have otherwise no guaranties that the processing to be awaiting for did start
```

Comment:
could you explain that?

Reply:
Taken from Swift equivalent... we may have an issue.

Reply:
Easy!

await is basically calling to ElasticSearch::flush - allowing to await that the indexing is ready.

However, async world being messy, you can very well end up flushing BEFORE the ES indexing even start. In that case await would return instantly while no message had been indexed.

Well, appart implementing `synchronous mailbox listeners` something that I recorded we did not wanted, I barely see other solutions.